Feature/sys annotate (#199)
cisco/mindmeld: source/userguide/auto_annotator.rst

Working with the Auto Annotator
=====================================================

The Auto Annotator

  - is a tool to automatically annotate or unannotate select entities across all labelled data in an application.
  - supports the development of custom Annotators.

.. note::

   The examples in this section require the :doc:`HR Assistant <../blueprints/hr_assistant>` blueprint application. To get the app, open a terminal and run ``mindmeld blueprint hr_assistant``.

.. warning::

   Changes by an Auto Annotator cannot be undone and Mindmeld does not backup query data. We recommend using version control software such as Github.

> 

Reviewer: We should add a Quick Start section telling how to use these commands directly:
```
mindmeld annotate --app-path "hr_assistant"
mindmeld unannotate --app-path "hr_assistant"

Alternatively
.. code-block:: python
	from mindmeld.auto_annotator import SpacyAnnotator 
	sa = SpacyAnnotator(app_path="hr_assistant")
	sa.annotate()

Author: Added a Quick Start section

		"""

		def __init__(self, app_path, config=None):
			super().__init__(app_path=app_path, config=config)
			
			# Add additional attributes if needed

		def parse(self, sentence, entity_types=None):
			""" 
			Args:
				sentence (str): Sentence to detect entities.
				entity_types (list): List of entity types to parse. If None, all
					possible entity types will be parsed.
			Returns: entities (list): List of entity dictionaries.
			"""

			# TODO: Add custom parse logic

			return entities

		def valid_entity_check(self, entity):
			""" 
			Args: entity (str): Name of entity to annotate.
			Returns: bool: Whether entity is valid.
			"""
			entity = entity.lower().strip()
			supported_entities = [

			# TODO: Add list of supported entity names
			
			]
			return entity in supported_entities

Entities returned by :attr:`parse()` must have the following format:

.. code-block:: python

	entity = { 
		"body": (substring of sentence), 
		"start": (start index), 
		"end": (end index + 1), 
		"dim": (entity type), 
		"value": (resolved value, if it exists), 
	}

Registering Annotator to Enable command-line Use
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> In order to use our :class:`CustomAnnotator` when running annotation/unannotation commands from the command-line, we must add the following line to the bottom of the :attr:`auto_annotator.py` file to register the class and ensure that it loads in the :attr:`cli.py` file.

Reviewer: Ideally for a person to implement their own custom annotator, they should not add to MindMeld's source code. Thoughts?

Author: That's a good point. We could make the cli command exclusive to the mindmeld inbuilt SpacyAnnotator and potential future mindmeld Annotators.

Developers can run their annotators in the same file with a small addition to the boiler plate code presented in the docs.

if __name__ == "__main__":
    custom_annotator = CustomAnnotator(app_path="hr_assistant")
    custom_annotator.annotate()

Thoughts on this approach?

Reviewer: Yup I think that should work

	Be sure to use the regex "or" (:attr:`|`) if applying rules at the same level of specificity. Otherwise, if written as separate rules, the latter will overwrite the former.

.. warning::
	By default, all files in all intents across all domains will be annotated with all supported entities. Before annotating consider including custom annotation rules in :attr:`config.py`. 

``'unannotate_supported_entities_only'`` (:class:`boolean`): By default, when the unannotate command is used only entities that the Annotator can annotate will be eligible for removal. 

``'unannotate'`` (:class:`list`): List of annotation rules in the same format as those used for annotation. These rules specify which entities should have their annotations removed. By default, :attr:`files` is None.

``'spacy_model'`` (:class:`str`): :attr:`en_core_web_lg` is used by default for the best performance. Alternative options are :attr:`en_core_web_sm` and :attr:`en_core_web_md`. This parameter is optional and is specific to the use of the :class:`SpacyAnnotator`.

> If the selected English model is not in the current environment it will automatically be downloaded. Refer to Spacy's documentation to learn more about their `English models <https://spacy.io/models/en>`_.

Reviewer: You can add a note about non-English models here.

The functionality for annotating and unannotating files is contained in :class:`Annotator` itself.
A developer simply needs to implement two methods to create a custom annotator.

Custom Annotator Boilerplate Code
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
This section includes boilerplate code to build a :class:`CustomAnnotator` class to which you can add to your own python file, let's call it :attr:`custom_annotator.py`
There are two "TODO"s. To implement a :class:`CustomAnnotator` class a developer has to implement the :meth:`parse` and :meth:`valid_entity_check` methods.

.. code-block:: python

	class CustomAnnotator(Annotator):
		""" Custom Annotator class used to generate annotations.
		"""

		def __init__(self, app_path, config=None):
			super().__init__(app_path=app_path, config=config)
			
			# Add additional attributes if needed

		def parse(self, sentence, entity_types=None):
			""" 
			Args:
				sentence (str): Sentence to detect entities.
				entity_types (list): List of entity types to parse. If None, all
					possible entity types will be parsed.
			Returns: entities (list): List of entity dictionaries.
			"""

			# TODO: Add custom parse logic

			return entities

> 		def valid_entity_check(self, entity):

Reviewer: Update this to `supported_entity_types`

+------------------------+-------------------------+-----------------------------------------------------------------------------+
| "sys_event"            | No                      | Named hurricanes, battles, wars, sports events, etc.                        |
+------------------------+-------------------------+-----------------------------------------------------------------------------+
| "sys_law"              | No                      | Named documents made into laws.                                             |
+------------------------+-------------------------+-----------------------------------------------------------------------------+
| "sys_langauge"         | No                      | Any named language.                                                         |
+------------------------+-------------------------+-----------------------------------------------------------------------------+
| "sys_work-of-art"      | No                      | Titles of books, songs, etc.                                                |
+------------------------+-------------------------+-----------------------------------------------------------------------------+

> | "sys_other-quantity"   | No                      | “40 degrees”                                                                |

Author: I have replaced "40 degrees" with "10 joules", "30 liters", "15 tons". These new examples are not resolved by duckling.

+------------------------+-------------------------+-----------------------------------------------------------------------------+

To detect entities in a single sentence first create an instance of the :class:`SpacyAnnotator` class.

.. code-block:: python

	from mindmeld.auto_annotator import SpacyAnnotator 

> 	sa = SpacyAnnotator(app_path=”hr_assistant”, model=”en_core_web_lg”)

Author: Yes, Annotator specific configurations are added through the config. Updated the docs to reflect this. Good catch!

Three entities are automatically recognized and a list of dictionaries is returned. Each dictionary represents a detected entity.:

.. code-block:: python
	
	[
		{
			'body': 'Apple',
			'start': 0,
			'end': 5,
			'value': {'value': 'Apple'},
			'dim': 'sys_org'
		},
		{
			'body': '$10',
			'start': 20,
			'end': 23,
			'value': {'value': 10, 'type': 'value', 'unit': '$'},
			'dim': 'sys_amount-of-money'
		},
		{
			'body': 'last monday',
			'start': 24,
			'end': 35,
			'value': {'value': '2020-09-21T00:00:00.000-07:00',
			'grain': 'day',
			'type': 'value'},
			'dim': 'sys_time'
		}
	]

The Auto Annotator detected "Apple" as :attr:`sys_org`. Moreover, it recognized "$10" as :attr:`sys_amount-of-money` and resolved its :attr:`value` as 10 and :attr:`unit` as "$".
Lastly, it recognized "last monday" as :attr:`sys_time` and resolved its :attr:`value` to be a timestamp representing the last monday from the current date.

In general, detected entities will be represented in the following format:

.. code-block:: python

	entity = { 

		“body”: (substring of sentence), 

		“start”: (start index), 

		“end”: (end index + 1), 

		“dim”: (entity type), 

> 		“value”: (resolved value, if it exists), 

Reviewer 2: ```suggestion
		"value": (resolved value, if it exists), 
```

For example, we can restrict the output of the previous example by doing the following:

.. code-block:: python
	
	allowed_entites = ["sys_org”, “sys_amount-of-money”, “sys_time”]
	sentence = "Apple stock went up $10 last monday."
	sa.parse(sentence=sentence, entity_types=allowed_entities) 

Auto Annotator Configuration
----------------------------

The :attr:`DEFAULT_AUTO_ANNOTATOR_CONFIG` shown below is the default config for an Annotator.
A custom config can be included in :attr:`config.py` by duplicating the default config and renaming it to :attr:`AUTO_ANNOTATOR_CONFIG`.
Alternatively, a custom config dictionary can be passed in directly to :class:`SpacyAnnotator` or any Annotator class upon instantiation.

.. code-block:: python

	DEFAULT_AUTO_ANNOTATOR_CONFIG = { 

		"annotator_class": "SpacyAnnotator",

		"overwrite": False, 

		"annotate": [ 

			{ 

				"domains": "*", 

				"intents": "*", 

				"files": "*", 

				"entities": "*", 

			} 

		], 

		"unannotate_supported_entities_only": True, 

> 		"unannotate": None, 

Author: Updated to consistently be single space

	}

Let's take a look at the allowed values for each setting in an Auto Annotator configuration.

``'annotator_class'`` (:class:`str`): The class in auto_annotator.py to use for annotation. By default, :class:`SpacyAnnotator` is used. 

``'overwrite'`` (:class:`bool`): Whether new annotations should overwrite existing annotations in the case of a span conflict. False by default. 

``'annotate'`` (:class:`list`): A list of annotation rules where each rule is represented as a dictionary. Each rule must have four keys: :attr:`domains`, :attr:`intents`, :attr:`files`, and :attr:`entities`.
Annotation rules are combined internally to create Regex patterns to match selected files. :attr:`*` can be used if all possibilities in a section are to be selected. :attr:`|` is used to represent “or”.  
Let's take a look at an example annotation rule:

.. code-block:: python

	{

		"domains": "faq|salary", 

		"intents": "*", 

		"files": "train.txt|test.txt", 

		"entities": "sys_amount-of-money|sys_time", 

	}

The rule above would annotate all text files named "train" or "test" in the "faq" and "salary" domains. Only sys_amount-of-money and sys_time entities would be annotated.

.. note::

> 	The order of the annotation rules matters. A rule will overwrite a previous rule if there are conflicts.

Author: Yes, since the overwrite in the default config refers to whether existing annotations should be overwritten

> 	The order of the annotation rules matters. A rule will overwrite a previous rule if there are conflicts.

Reviewer 2: Is it possible to reverse this behavior? In a system of ordered rules like this, you could have phantom blocking effects, e.g.
1. Rule 1 adds an entity: `{A B} C D`
2. Rule 2 replaces that entity with another `A {B C} D`
3. Rule 3 replaces that entity with another `A B {C D}`

The result is that Rule 1 fails to apply just because it conflicts with a rule that also didn't apply.

> 	The order of the annotation rules matters. A rule will overwrite a previous rule if there are conflicts.

Author: The motivation behind this design is to be able to annotate entities from general to specific rules.

For example, say you want to annotate entities in the "date" domain in the HR assistant with the general rule to annotate sys_time and sys_date. However, for the "get_date_range_aggregate" you would also like to include "sys_interval" then you would a second rule specifying this intent and the additional entity. In this manner, the previous rule isn't overwritten but expanded upon at a more specific level.

An alternative to avoid any conflict is to write rules for each intent, however, this may quickly become very verbose and time consuming.

``'unannotate_supported_entities_only'`` (:class:`boolean`): By default, when the unannotate command is used only entities that the Annotator can annotate will be eligible for removal. 

``'unannotate'`` (:class:`list`): List of annotation rules in the same format as those used for annotation. These rules specify which entities should have their annotations removed. By default, :attr:`files` is None.

``'spacy_model'`` (:class:`str`): :attr:`en_core_web_lg` is used by default for the best performance. Alternative options are :attr:`en_core_web_sm` and :attr:`en_core_web_md`. This parameter is optional and is specific to the use of the :class:`SpacyAnnotator`.
If the selected english model is not in the current environment it will automatically be downloaded. Refer to Spacy's documentation to learn more about their `english models <https://spacy.io/models/en>`_.

Using the Auto Annotator
------------------------

The Auto Annotator can be used by importing a class that implements the :class:`Annotator` abstract class in the :mod:`auto_annotator` module or through the command-line.

> We will demonstrate both approaches for Unannotation and Annotation using the :class:`SpacyAnnotator` class.

Author: That's correct, a more natural workflow is to annotate and then unannotate. While that makes more sense in practice, since blueprints are already annotated it's easier to demonstrate in this order. I have added this sentence to help clarify what the natural workflow is: 

"For the this tutorial we will first describe unannotation and then annotation, however, in a normal workflow you are likely to annotate first and then unannotate any annotations you are not pleased with."

Annotate
^^^^^^^^

To annotate by creating an instance of the :class:`Annotator` class, run:

.. code-block:: python

	from mindmeld.auto_annotator import SpacyAnnotator 
	sa = SpacyAnnotator(app_path="hr_assistant")

	sa.annotate()

Alternatively, you can :attr:`annotate` using the Command-Line:

.. code-block:: console

	mindmeld annotate --app-path "hr_assistant"

Let's annotate :attr:`sys_person` entities from the :attr:`get_hierarchy_up` intent in the :attr:`hierarchy` domain.
To do this, we can add the following :attr:`AUTO_ANNOTATOR_CONFIG` dictionary to :attr:`config.py`.
Notice that we are setting :attr:`overwrite` to True since we want to replace the existing custom entity label, :attr:`name`.

.. code-block:: python

	AUTO_ANNOTATOR_CONFIG = { 

		"annotator_class": "SpacyAnnotator",
		"overwrite": True, 
		"annotate": [
			{ 
				"domains": "hierarchy", 
				"intents": "get_hierarchy_up", 
				"files": "train.txt",
				"entities": "sys_person", 
			}
		],
		"unannotate_supported_entities_only": True, 
		"unannotate": None
	}

Before running the annotation, we can see the first four queries in the train.txt file for the :attr:`get_hierarchy_up` intent: 

> .. code-block:: python

Reviewer 2: We don't need any syntax highlighting for query data. 
```suggestion
.. code-block:: none
```

Working with the Auto Annotator
=====================================================

The Auto Annotator

  - is a tool to automatically annotate or unannotate select entities across all labelled data in an application.
  - supports the development of custom Annotators.

.. note::

   The examples in this section require the :doc:`HR Assistant <../blueprints/hr_assistant>` blueprint application. To get the app, open a terminal and run ``mindmeld blueprint hr_assistant``.

.. warning::

   Changes by an Auto Annotator cannot be undone and Mindmeld does not backup query data. We recommend using version control software such as Github.

Quick Start
-----------

> This section briefly explains the use of the unannotate and annotate commands. For more details, read the next section.

Reviewer: ```suggestion
This section briefly explains the use of the ``annotate`` and ``unannotate`` commands. For more details, read the next section.
```

> First, copy and paste the config below into the :attr:`config.py` of the mindmeld app that you would like to remove annotations.

Reviewer: ```suggestion
First, copy and paste the configuration below into the :attr:`config.py` of the MindMeld app that you would like to operate on.
```

.. code-block:: python

	AUTO_ANNOTATOR_CONFIG = { 

		"annotator_class": "SpacyAnnotator",
		"overwrite": False, 
		"annotate": [{"domains": ".*", "intents": ".*", "files": ".*", "entities": ".*"}],
		"unannotate_supported_entities_only": True, 
		"unannotate": [{"domains": ".*", "intents": ".*", "files": ".*", "entities": ".*"}],
	}

> To unannotate you can run python code or use the command-line.

Reviewer: ```suggestion
To "unannotate" the files, you can run Python code or use the command-line.
```

Reviewer: Can you explain the `unannotate` command here more? Why do we recommend people to `unannotate` first before running `annotate`?

Author: In a typical workflow, you would ofcourse `annotate` and then `unannotate`. Since a blueprint with existing annotations is being used, it seems easier to demonstrate unannotation first and then demonstrate annotation on unannotated text.

If this seems confusing, I can alternatively demonstrate annotation first and have users set "overwrite" to True and then demonstrate unannotating the all or a subset of the newly added annotations? This actually may better represent the real-world use case of this feature.

Python:

.. code-block:: python

	from mindmeld.auto_annotator import SpacyAnnotator 
	sa = SpacyAnnotator(app_path="hr_assistant")

	sa.unannotate()

> Command-line:

Reviewer: Let's show the Python command first

Author: The Python command for "annotate" and "unannotate" is shown before the CLI command

Unannotate
- Python
- CLI

Annotate
- Python
- CLI

Is there a different structure that you would prefer?

Using the Auto Annotator
------------------------

The Auto Annotator can be used by importing a class that implements the :class:`Annotator` abstract class in the :mod:`auto_annotator` module or through the command-line.
We will demonstrate both approaches for unannotation and annotation using the :class:`SpacyAnnotator` class. We will first describe unannotation and then annotation. However, in a normal workflow you are likely to annotate first and then unannotate any annotations you are not pleased with.

Unannotate
^^^^^^^^^^
To unannotate by creating an instance of the :class:`Annotator` class, run:

.. code-block:: python

	from mindmeld.auto_annotator import SpacyAnnotator 
	sa = SpacyAnnotator(app_path="hr_assistant")

	sa.unannotate()

Alternatively, you can :attr:`unannotate` using the command-line:

.. code-block:: console

	mindmeld unannotate --app-path "hr_assistant"

If you see the following message, you need to update the unannotate parameter in your custom :attr:`AUTO_ANNOTATOR_CONFIG` dictionary in :attr:`config.py`. You can refer to the config specifications in the section above.

.. code-block:: console

> 	'unannotate' is None in the config. Nothing to unannotate.

Reviewer: We probably should update this error to be clearer?

```suggestion
	'unannotate' field is not configured or misconfigured in the `config.py`. We can't find any file to unannotate.
```

Let's unannotate :attr:`sys_time` entities from the :attr:`get_date_range_aggregate` intent in the :attr:`date` domain.
To do this, we can add the following :attr:`AUTO_ANNOTATOR_CONFIG` dictionary to :attr:`config.py`.

.. code-block:: python

	AUTO_ANNOTATOR_CONFIG = { 

		"annotator_class": "SpacyAnnotator",
		"overwrite": False, 
		"annotate": [{"domains": ".*", "intents": ".*", "files": ".*", "entities": ".*"}],
		"unannotate_supported_entities_only": True, 
		"unannotate": [
			{ 
				"domains": "date", 
				"intents": "get_date_range_aggregate", 
				"files": "train.txt",
				"entities": "sys_time", 
			}
		], 
	}

.. note::

> 	The content of :attr:`annotate` in the config has no effect on unannotation. Similarly, :attr:`unannotate` in the config has no affect on annotation. These processes are independent and are only affected by the corresponding parameter in the config.

Reviewer: What is the reason for having separate params for `annotate` and `unannotate`? should they be operating on the same files?

Author: It may not always be the case that you would want to unannotate and annotate the same entities. Often times you would like to annotate a set of entities and then unannotate a subset that you realised you would rather not have.

By having these as separate parameters, a developer doesn't have to copy and paste a different set of rules each time they switch methods and doesn't risk accidentally unannotating their existing annotations

.. code-block:: python

	DEFAULT_AUTO_ANNOTATOR_CONFIG = { 

		"annotator_class": "SpacyAnnotator",
		"overwrite": False, 
		"annotate": [ 
			{ 
				"domains": ".*", 
				"intents": ".*", 
				"files": ".*", 
				"entities": ".*", 
			} 
		], 
		"unannotate_supported_entities_only": True, 
		"unannotate": None, 
	}

Let's take a look at the allowed values for each setting in an Auto Annotator configuration.

``'annotator_class'`` (:class:`str`): The class in auto_annotator.py to use for annotation when invoked from the command line. By default, :class:`SpacyAnnotator` is used. 

``'overwrite'`` (:class:`bool`): Whether new annotations should overwrite existing annotations in the case of a span conflict. False by default. 

``'annotate'`` (:class:`list`): A list of annotation rules where each rule is represented as a dictionary. Each rule must have four keys: :attr:`domains`, :attr:`intents`, :attr:`files`, and :attr:`entities`.
Annotation rules are combined internally to create Regex patterns to match selected files. The character :attr:`.*` can be used if all possibilities in a section are to be selected, while possibilities within

> a section are expressed with the usual Regex special characters, such as :attr:` . ` for any single character and :attr:`|` to represent "or". 

Reviewer: I think the `.` character is not rendered correctly
![image](https://user-images.githubusercontent.com/1332283/95546287-0571b280-09b5-11eb-9735-52e1c064c849.png)

If the selected English model is not in the current environment it will automatically be downloaded. Refer to Spacy's documentation to learn more about their `English models <https://spacy.io/models/en>`_.

Creating a Custom Annotator
---------------------------
The :class:`SpacyAnnotator` is a subclass of the abstract base class :class:`Annotator`.
The functionality for annotating and unannotating files is contained in :class:`Annotator` itself.
A developer simply needs to implement two methods to create a custom annotator.

Custom Annotator Boilerplate Code
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
This section includes boilerplate code to build a :class:`CustomAnnotator` class to which you can add to your own python file, let's call it :attr:`custom_annotator.py`
There are two "TODO"s. To implement a :class:`CustomAnnotator` class a developer has to implement the :meth:`parse` and :meth:`valid_entity_check` methods.

.. code-block:: python

	class CustomAnnotator(Annotator):
		""" Custom Annotator class used to generate annotations.
		"""

		def __init__(self, app_path, config=None):
			super().__init__(app_path=app_path, config=config)
			
			# Add additional attributes if needed

> 		def parse(self, sentence, entity_types=None):

Reviewer: where do the entity types come from? probably we should make it clear that these entity types are the ones we want to annotate

Author: Updating the docs for this function:

```
        Args:
            sentence (str): Sentence to detect entities.
            entity_types (list): List of entity types to annotate. If None, all
                possible entity types will be annotated.
```

			""" 
			Args:
				sentence (str): Sentence to detect entities.
				entity_types (list): List of entity types to parse. If None, all
					possible entity types will be parsed.
			Returns: entities (list): List of entity dictionaries.
			"""

			# TODO: Add custom parse logic

			return entities

		def valid_entity_check(self, entity):
			""" 
			Args: entity (str): Name of entity to annotate.
			Returns: bool: Whether entity is valid.
			"""
			entity = entity.lower().strip()

> 			supported_entities = [

Reviewer: we should make this into a class/instance variable

Author: Does this happen automatically if you use @property?

Author: Could we do the following in the class init as an alternative to using @ property? 

`self.supported_entity_types = self.get_supported_entity_types()`

			# TODO: Add list of supported entity names
			
			]
			return entity in supported_entities
	
	if __name__ == "__main__":
		custom_annotator = CustomAnnotator(app_path="hr_assistant")
		custom_annotator.annotate()

Entities returned by :attr:`parse()` must have the following format:

.. code-block:: python

	entity = { 
		"body": (substring of sentence), 
		"start": (start index), 
		"end": (end index + 1), 
		"dim": (entity type), 
		"value": (resolved value, if it exists), 
	}

To run your custom Annotator, simply run in the command line: :attr:`python custom_annotator.py`.
To run unannotation with your custom Annotator, change the last line in your script to :attr:`custom_annotator.unannotate()`.

Getting Custom Parameters from the Config
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

:attr:`spacy_model` is an example of an optional parameter in the config that is relevant only for a specific :class:`Annotator` class.

.. code-block:: python

	AUTO_ANNOTATOR_CONFIG = { 
		... 

> 		"spacy_model": "en_core_web_md",

Author: I haven't yet tried using another language model but I have seen on Spacy's website that the output of NER for models that support it uses the same interface, so it should likely work so long as duckling also supports that language as well

A developer can't use a non-english model currently since an exception would be thrown if the model is not in this list: [EN_CORE_WEB_SM, EN_CORE_WEB_MD, EN_CORE_WEB_LG]

> 		"spacy_model": "en_core_web_md",

Reviewer: I see - can you run a quick test to see if that can work with non-english language? in that case, we should allow support for non-english language, but just add a warning if it's not in the list.

> 		"spacy_model": "en_core_web_md",

Author: I was able to load up a Spacy Spanish annotator using es_core_news_lg. I had it parse the following text:

`text = "El domingo, el Gobierno de Navarra, que ha registrado 675,3 casos por cada 100.000 habitantes en las dos últimas semanas, ordenó que todos los bares y restaurantes reduzcan su capacidad al 30% y cierren a las 22.00 horas."`

It was only able to identify "Gobierno de Navarra" as location and missed the numbers and time-related phrases. When I passed the same text into duckling (language="es"), duckling is able to identify many more entities (but not able to collapse candidates to a final set of course).

Will go ahead and add the support and warning as suggested

> 		"spacy_model": "en_core_web_md",

Author: Done

		... 
	}

:class:`SpacyAnnotator` checks if :attr:`spacy_model` exists in the config, and if it doesn't, it will use the default value of "en_core_web_lg".

> .. code-block:: python

Author: Removed
